src/core_ocean/shared/mpas_ocn_gm.F

```diff
@@ -196,323 +195,204 @@ subroutine ocn_gm_compute_Bolus_velocity(diagnosticsPool, meshPool, scratchPool)
-            relativeSlopeTopOfEdge(k,iEdge) = - gradDensityTopOfEdge(k,iEdge) / min(dDensityDzTopOfEdge(k,iEdge),-epsGM)
-         end do
+         !$omp do schedule(runtime) private(maxLocation, cell1, cell2, k, BruntVaisalaFreqTopEdge, maxN)
+         do iCell = 1,nCells
+            k = min(maxLevelCell(iCell)-1,max(1,indMLD(iCell)))
+            k=1
+            maxN =  max(BruntVaisalaFreqTop(k,iCell), 0.0_RKIND)
+            do while (BruntVaisalaFreqTop(k+1,iCell) > maxN .and. k < maxLevelCell(iCell)-1)
+              k = k + 1
+              maxN = max(BruntVaisalaFreqTop(k,iCell), 0.0_RKIND)
+            enddo
 
-         ! Since dDensityDzTopOfEdge is guaranteed to be zero on the top surface, relativeSlopeTopOfEdge on the top
-         ! surface is identified with its value on the second interface.
-         relativeSlopeTopOfEdge(1,iEdge) = relativeSlopeTopOfEdge(2,iEdge)
+            maxLocation = k 
+            do k=maxLocation,maxLevelCell(iCell)
+               BruntVaisalaFreqTopEdge = max(BruntVaisalaFreqTop(k,iCell), 0.0_RKIND)
 
-         ! dDensityDzTopOfEdge may or may not equal zero on the bottom surface, depending on whether
-         ! maxLevelEdgeTop(iEdge) = maxLevelEdgeBottom(iEdge). But here we
-         ! take a simplistic approach and identify relativeSlopeTopOfEdge on the bottom surface with its value on
-         ! the interface just above.
-         relativeSlopeTopOfEdge( maxLevelEdgeTop(iEdge)+1, iEdge ) = relativeSlopeTopOfEdge( max(1,maxLevelEdgeTop(iEdge)), iEdge )
+               kappaGMCell(k,iCell) = min(max(config_GM_min_stratification_ratio, &
+                                          BruntVaisalaFreqTopEdge / (maxN + 1.0E-10_RKIND)), &
+                                          1.0_RKIND)
+            enddo
+         enddo
+         !$omp end do
+      endif
 
-      end do
-      !$omp end do
+      if (config_Redi_use_surface_taper) then
+        !$omp do schedule (runtime) private(zMLD, k)
+        do iCell = 1,nCells
+          k = max(1,indMLD(iCell))
+          zMLD = ssh(iCell) - zMid(k,iCell)
 
-      nEdges = nEdgesArray( 3 )
+          do k = 1, indMLD(iCell)
+             kappaRediSfcTaper(k,iCell) = abs((ssh(iCell) - zMid(k,iCell)) / (zMLD))
+           enddo
+         enddo
+        !omp end do
+      endif
 
-      ! slope can be unbounded in regions of neutral stability, reset to the large, but bounded, value
-      ! values is hardwrite to 1.0, this is equivalent to a slope of 45 degrees
-      !$omp do schedule(runtime) private(k)
-      do iEdge = 1, nEdges
-         do k = 1, nVertLevels
-            relativeSlopeTopOfEdge(k, iEdge) = max( min( relativeSlopeTopOfEdge(k, iEdge), 1.0_RKIND), -1.0_RKIND)
+      nCells = nCellsArray( 3 )
+      !$omp do schedule(runtime) private(invAreaCell, iEdge, cell1, cell2, k, iCellSelf) &
+      !$omp& private(dcEdgeInv, areaEdge, drhoDT, drhoDS, dTdx, dSdx, drhoDx) &
+      !$omp& private(sfcTaper, sfcTaperUp, sfcTaperDown, sfcTaperFactor)
```

Comment:
@philipwjones, @mattdturner or @amametjanov: I am failing a test for bfb match between 1 and 2 threads. Could you please check the above three lines? I added all variables that need to be thread-private. I have four allocated column arrays: 
```dzTop, dTdzTop, dSdzTop, k33Norm```
Do I put those in `omp private`? I tried, but got a compile error.

If you could check for any omp directive errors in this file, that would be helpful.

Reply:
It looks like `iCellSelf` is being initialized inside the loop but is not declared `private`.

Reply:
Please try with this update that has additional private variables:
```
azamatm@thetalogin2:~/repos/E3SM-clone/components/mpas-source> git diff
diff --git a/src/core_ocean/shared/mpas_ocn_gm.F b/src/core_ocean/shared/mpas_ocn_gm.F
index 2490cf07..2d152190 100644
--- a/src/core_ocean/shared/mpas_ocn_gm.F
+++ b/src/core_ocean/shared/mpas_ocn_gm.F
@@ -270,9 +270,9 @@ contains
       endif
 
       nCells = nCellsArray( 3 )
-      !$omp do schedule(runtime) private(invAreaCell, iEdge, cell1, cell2, k, iCellSelf) &
-      !$omp& private(dcEdgeInv, areaEdge, drhoDT, drhoDS, dTdx, dSdx, drhoDx) &
-      !$omp& private(sfcTaper, sfcTaperUp, sfcTaperDown, sfcTaperFactor)
+      !$omp do schedule(runtime) private(invAreaCell, iEdge, cell1, cell2, k, iCellSelf, &
+      !$omp& dcEdgeInv, areaEdge, drhoDT, drhoDS, dTdx, dSdx, drhoDx, &
+      !$omp& sfcTaper, sfcTaperUp, sfcTaperDown, sfcTaperFactor, i, slopeTaperDown, slopeTaperUp)
       do iCell = 1, nCells
          invAreaCell = 1.0_RKIND/areaCell(iCell)
          k33(1:maxLevelCell(iCell)+1,iCell) = 0.0_RKIND
@@ -376,8 +376,8 @@ contains
                slopeTriadDown(k,iCellSelf,iEdge) = &
                   slopeTaperDown * sfcTaperDown * slopeTriadDown(k,iCellSelf,iEdge)
 
-            end do
-        end do
+            end do ! maxLevelEdgeTop(iEdge)
+        end do ! nEdgesOnCell(iCell)
 
          ! Normalize k33
         do k = 2, maxLevelCell(iCell)
```